### DISCUSSION TITLE
Making the C stack invisible to bytecode by inserting shim frames

### DISCUSSION THREAD

**Discussion by markshannon** [faster-cpython/ideas#391] [Ideas]
### Motivation

The `RETURN_VALUE` instruction needs check to whether the current frame is an "entry" frame (it was the frame that was passed to `PyEval_EvalFrameDefault`) and either do an "internal" return (pop the frame, and continue in the caller frame) or an "external" return (do a return in the C code).

This isn't a big deal deal for `RETURN_VALUE`, but could get cumbersome if we want to do the same for `YIELD_VALUE`, to allow 
effective specialization of `FOR_ITER` for generators, and specialization of `SEND` for coroutines.

By inserting a shim frame on entry to `PyEval_EvalFrameDefault` and in `gen.send` we can shield `RETURN_VALUE` and `YIELD_VALUE` from knowing anything about the C stack. 

### Shim frame

A shim frame is a `_PyInterpreterFrame` inserted into the stack on entry to `PyEval_EvalFrameDefault`. The code object for a shim frame would consist of the single instruction `EXIT`, which would perform a C return.

### Cost/benefit

Every call to `PyEval_EvalFrameDefault` would suffer additional cost of pushing a `_PyInterpreterFrame`, whereas every `RETURN_VALUE` would only save the cost of a single branch. Unless the number of `RETURN_VALUE`s is a *lot* more than the number of calls to `PyEval_EvalFrameDefault` then it clearly isn't worth doing at the moment.

However, once we want to specialize `YIELD_VALUE`, then this starts to make a lot more sense, as we can implement `gen.send` in bytecode that converts a yield into a return and wraps a return value in a `StopIteration`
```
def gen_send(coro, value):
      LOAD_FAST coro
      LOAD_FAST value
      SEND return_label
      RETURN_VALUE # yield becomes return
return_label:
     # Might need to explicitly clear the coroutine here.
     LOAD_EXCEPTION StopIteration
     PUSH_NULL
     SWAP 3
     CALL 1
     RAISE_VARARGS 1 # return becomes raise StopIteration
```

### Specializations of `FOR_ITER` and `SEND`

If we add another "return address" to `_PyInterpreterFrame` (or just an offset from the current `prev_instr`) then `YIELD_VALUE` can jump to that, much as `RETURN_VALUE` jumps to `prev_instr`. 

The specialization of `FOR_ITER` for generators would set this additional "return address", push the generator's frame and jump into that frame.
The specialization of `SEND` for coroutines would do much the same, as `FOR_ITER` is basically `LOAD_CONST None; SEND`.

**Comment by brandtbucher** [comment]
Are you planning on working on this?

If not, I’d like to take it. I’m going to be picking the `yield from`/`await` stuff back up today, and I also want to use these shim frames for inlining pure-Python binary operators, etc.

**Reply by markshannon** [reply]
Go ahead.

I'm working on related stuff, making the frame popping and pushing a bit slicker: https://github.com/faster-cpython/ideas/issues/111#issuecomment-1100903897, so there might be some merge conflicts. 

https://github.com/python/cpython/pull/92960 (done)
https://github.com/python/cpython/pull/92957 (not urgent)
